File: common/src/main/java/cn/breadnicecat/candycraftce/block/blocks/CandySaplingBlock.java
package cn.breadnicecat.candycraftce.block.blocks;

import cn.breadnicecat.candycraftce.block.CBlockTags;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.SaplingBlock;
import net.minecraft.world.level.block.grower.AbstractTreeGrower;
import net.minecraft.world.level.block.state.BlockState;

/**
 * Created in 2024/2/6
 * Project: candycraftce
 *
 * @author <a href="https://github.com/Bread-NiceCat">Bread_NiceCat</a>
 * <p>
 */
public class CandySaplingBlock extends SaplingBlock implements ISugarTarget {
	protected CandySaplingBlock(AbstractTreeGrower abstractTreeGrower, Properties properties) {
		super(abstractTreeGrower, properties);
	}
	
	
	@Override
	protected boolean mayPlaceOn(BlockState state, BlockGetter level, BlockPos pos) {
		return state.is(CBlockTags.BT_CANDY_PLANT_SUITABLE);
	}
	
	@Override
	public boolean isValidSugarTarget(LevelReader level, BlockPos pos, BlockState state, boolean isClient) {
		return true;
	}
	
	@Override
	public boolean isSugarSuccess(ServerLevel level, RandomSource rand, BlockPos pos, BlockState state) {
		return rand.nextFloat() < 0.45;
	}
	
	@Override
	public void performSugar(ServerLevel level, RandomSource rand, BlockPos pos, BlockState state) {
		advanceTree(level, pos, state, rand);
	}
	
	
	@Override
	public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state, boolean isClient) {
		return false;
	}
	
	@Override
	public boolean isBonemealSuccess(Level level, RandomSource random, BlockPos pos, BlockState state) {
		return false;
	}
	
	@Override
	public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/block/blocks/plant/CandySaplingBlock.kt
package cn.breadnicecat.candycraftce.core.block.blocks.plant

import cn.breadnicecat.candycraftce.core.tag.CTags
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.SaplingBlock
import net.minecraft.world.level.block.grower.AbstractTreeGrower
import net.minecraft.world.level.block.state.BlockState

/**
 * Created by NiceCat on 2026/1/9.
 * Project: candycraftce
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 *
 */
class CandySaplingBlock(treeGrower: AbstractTreeGrower, properties: Properties) :
    SaplingBlock(treeGrower, properties), ISugarTarget {
    override fun mayPlaceOn(state: BlockState, level: BlockGetter, pos: BlockPos): Boolean {
        return level.getBlockState(pos).`is`(CTags.CBlockTags.candy_plant_suitable)
    }

    //去骨粉
    override fun isValidBonemealTarget(
        level: LevelReader,
        pos: BlockPos,
        state: BlockState,
        isClient: Boolean,
    ): Boolean = false

    override fun isBonemealSuccess(
        level: Level,
        random: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ): Boolean = false

    override fun performBonemeal(
        level: ServerLevel,
        random: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ) {
    }

    override fun isValidSugarTarget(
        level: LevelReader,
        pos: BlockPos,
        state: BlockState,
        isClient: Boolean,
    ): Boolean = true

    override fun isSugarSuccess(
        level: ServerLevel,
        rand: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ): Boolean = level.random.nextFloat() < 0.45

    override fun performSugar(
        level: ServerLevel,
        rand: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ) = this.advanceTree(level, pos, state, rand)
}